### PR TITLE
SCL-18372 Support SBT extension in BSP import

### PR DESCRIPTION
**Pull request by retronym**
Tested against: https://github.com/sbt/sbt/pull/6553

### CODE REVIEW

**Comment by retronym** [comment]
image

image

image

File: .idea/sbt.xml [hunk 1]
```diff
@@ -15,26 +15,38 @@
...
             <option value="$PROJECT_DIR$/scala/uast" />
             <option value="$PROJECT_DIR$/scala/worksheet" />
-            <option value="$PROJECT_DIR$/target/tools/idea-runner" />
+            <option value="$PROJECT_DIR$/scala/worksheet-repl-interface" />
             <option value="$PROJECT_DIR$/target/tools/zinc" />
+            <option value="$PROJECT_DIR$/tasty/runtime" />
           </set>
         </option>
-        <option name="sbtVersion" value="0.13.18" />
+        <option name="sbtVersion" value="1.4.9" />
```
Comment (jastice) [inline .idea/sbt.xml]: did you change to this sbt version locally or how did this version come in?

**Review by jastice — APPROVED** [review]
Thanks so much! Is it still draft because the changes in sbt aren't through yet? If it works against your version and is backward compatible we can merge it.

File: .idea/sbt.xml (hunk 1 above)
Comment (retronym) [reply .idea/sbt.xml]: I didn't mean to check in any changes to this file. I guess it auto-updated when I re-imported the SBT project, maybe it is out of date? I'll remove it from this PR in any case.

**Comment by retronym** [comment]
> Is it still draft because the changes in sbt aren't through yet?

## `-build` naming convention

The main question I had was whether it was really still necessary for the BSP server to adhere to the `-build` naming convention for the SBT target.


There are comments suggesting that this is no longer the only, or desired, way to make the link. 

```scala
moduleByUri.orElse {
        // TODO remove in 2018.3+
        // this is the old way of finding a build module which breaks if the way the module name is assigned changes
        Option(manager.findModuleByName(module.getName + Sbt.BuildModuleSuffix))
      }.fold(DefinitionModule(module): TargetModule) { module =>
        SbtModuleWithScope(module, module.getModuleWithDependenciesAndLibrariesScope(false))
      }
```

But there seem to be plenty of other usages of the constant Sbt.BuildModuleSuffix (e.g in scala/scala-impl/src/org/jetbrains/plugins/scala/project/package.scala) which suggest that not all features work without the naming convention.

I guess that question could be answered later; for now I'll just keep the `-build` suffix.

## Source JARs not yet tested

The second thing I haven't tested yet is whether source JARs will be correctly attached. I am yet to implement that in the BSP server. Again, that need not block merging of this PR.

## UI changes?

You might want to extend the BSP settings UI to let user opt-out of the .sbt import. I guess some people find it annoying to have two versions of scala-library in the overall project. There could also be a UI option to control whether SBT sources/docs are downloaded. I'll leave those choices to you though.

## Testing?

I haven't included any tests for the new functionality. It's not a core feature so perhaps you're okay with that :)

File: bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala [hunk 2]
```diff
@@ -738,9 +754,10 @@ private[importing] object BspResolverLogic {
 
       case SbtModule(_, scalaSdkData, sbtData) =>
         val scalaSdkNode = new DataNode[ScalaSdkData](ScalaSdkData.Key, scalaSdkData, moduleNode)
-        val sbtNode = new DataNode[SbtBuildModuleDataBsp](SbtBuildModuleDataBsp.Key, sbtData, moduleNode)
+        val sbtBuildModuleDataNode = new DataNode[SbtBuildModuleDataBsp](SbtBuildModuleDataBsp.Key, sbtData, moduleNode)
         moduleNode.addChild(scalaSdkNode)
-        moduleNode.addChild(sbtNode)
+        moduleNode.addChild(sbtBuildModuleDataNode)
+        moduleNode.addChild(new SbtModuleNode(SbtModuleData(moduleData.id, new URI(moduleData.id))).toDataNode)
```
Comment (retronym) [inline bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala:760]: I didn't really know what I was doing in this part of the patch, so worth double checking if it makes sense.

**Comment by jastice** [comment]
> The main question I had was whether it was really still necessary for the BSP server to adhere to the -build naming convention for the SBT target.

It *shouldn't* be necessary, as on sbt side I replaced this logic with the `org.jetbrains.sbt.project.data.SbtBuildModuleData#buildFor` to allow a less error-prone build-module relationship. But you're right it's still used in a bunch of places, I should take care of that... Regardless, this logic is separate from how BSP imported modules behave.

> I haven't included any tests for the new functionality. It's not a core feature so perhaps you're okay with that :)

If it passes existing tests it's good enough for me right now. I'll revisit once an sbt version with build target support is released.

File: bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala (hunk 2 above)
Comment (jastice) [reply bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala:760]: oh yeah that's using a class from the sbt support, which it shouldn't in the BSP context :D

**Comment by retronym** [comment]
I agree, let's book the progress and we can do a round of follows if needed.

File: bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala (hunk 2 above)
Comment (retronym) [reply bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala:760]: Maybe there is an argument that its okay to have this dependency when the SBT extension is used. This lets us share the project settings UI.

Comment (jastice) [reply bsp/src/org/jetbrains/bsp/project/importing/BspResolverLogic.scala:760]: Okay, let's leave this redundancy in for now, though I hope to eliminate it when unifying BSP and sbt. My idea is to make BSP the default sbt mechanism for newer sbt versions, transparently to the user.

**Comment by jastice** [comment]
I had to update it a bit, but there we go!

**Comment by retronym** [comment]
@jastice SBT 1.5.5 is now released. It includes my patch to export the SBT target.

I tried updating the scala/scala SBT build to this version an importing with the latest plugin nightly (2021.2.400) and also with a locally built version of the plugin.

While the "root-build" project is imported, and some highlighting of the `./*.sbt` is working, it seems that the `root-build` module does not have the `./project/*.scala` files included as sources nor does it have the UI in the Project Structure dialog to show the SBT settings editor.

The persisted XML starts with:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<module external.linked.project.id="file:$MODULE_DIR$/../../#root-build" external.linked.project.path="$MODULE_DIR$/../.." external.root.project.path="$MODULE_DIR$/../.." external.system.id="BSP" sbt.imports="import _root_.scala.xml.{TopScope=&amp;gt;SUB:DOLLARscope}, import _root_.sbt._, import _root_.sbt.Keys._, import _root_.sbt.nio.Keys._, import _root_.sbt.ScriptedPlugin.autoImport._, _root_.sbt.plugins.JUnitXmlReportPlugin.autoImport._, _root_.sbt.plugins.MiniDependencyTreePlugin.autoImport._, _root_.com.typesafe.tools.mima.plugin.MimaPlugin.autoImport._, _root_.de.heikoseeberger.sbtheader.AutomateHeaderPlugin.autoImport._, _root_.de.heikoseeberger.sbtheader.HeaderPlugin.autoImport._, _root_.pl.project13.scala.sbt.JmhPlugin.autoImport._, _root_.sbtwhitesource.WhiteSourcePlugin.autoImport._, _root_.scala.build.BuildSettings.autoImport._, _root_.scala.build.MimaFilters.autoImport._, import _root_.sbt.plugins.IvyPlugin, _root_.sbt.plugins.JvmPlugin, _root_.sbt.plugins.CorePlugin, _root_.sbt.ScriptedPlugin, _root_.sbt.plugins.SbtPlugin, _root_.sbt.plugins.SemanticdbPlugin, _root_.sbt.plugins.JUnitXmlReportPlugin, _root_.sbt.plugins.Giter8TemplatePlugin, _root_.sbt.plugins.MiniDependencyTreePlugin, _root_.com.typesafe.tools.mima.plugin.MimaPlugin, _root_.de.heikoseeberger.sbtheader.AutomateHeaderPlugin, _root_.de.heikoseeberger.sbtheader.HeaderPlugin, _root_.pl.project13.scala.sbt.JmhPlugin, _root_.sbtwhitesource.WhiteSourcePlugin, _root_.scalabuild.JitWatchFilePlugin, _root_.scala.build.BuildSettings, _root_.scala.build.License, _root_.scala.build.MimaFilters, _root_.io.github.retronym.SbtArgsFilePlugin, import SUB:DOLLARa28383612b82af97ec2d.`publishDists`, import SUB:DOLLARa28383612b82af97ec2d.`bootstrap`, import SUB:DOLLARa28383612b82af97ec2d.`library`, import SUB:DOLLARa28383612b82af97ec2d.`reflect`, import SUB:DOLLARa28383612b82af97ec2d.`compilerOptionsExporter`, import SUB:DOLLARa28383612b82af97ec2d.`compiler`, import SUB:DOLLARa28383612b82af97ec2d.`interactive`, import SUB:DOLLARa28383612b82af97ec2d.`repl`, import SUB:DOLLARa28383612b82af97ec2d.`replJline`, import SUB:DOLLARa28383612b82af97ec2d.`replJlineEmbedded`, import SUB:DOLLARa28383612b82af97ec2d.`scaladoc`, import SUB:DOLLARa28383612b82af97ec2d.`scalap`, import SUB:DOLLARa28383612b82af97ec2d.`partest`, import SUB:DOLLARa28383612b82af97ec2d.`specLib`, import SUB:DOLLARa28383612b82af97ec2d.`bench`, import SUB:DOLLARa28383612b82af97ec2d.`junit`, import SUB:DOLLARa28383612b82af97ec2d.`scalacheck`, import SUB:DOLLARa28383612b82af97ec2d.`osgiTestFelix`, import SUB:DOLLARa28383612b82af97ec2d.`osgiTestEclipse`, import SUB:DOLLARa28383612b82af97ec2d.`partestJavaAgent`, import SUB:DOLLARa28383612b82af97ec2d.`test`, import SUB:DOLLARa28383612b82af97ec2d.`manual`, import SUB:DOLLARa28383612b82af97ec2d.`libraryAll`, import SUB:DOLLARa28383612b82af97ec2d.`scalaDist`, import SUB:DOLLARa28383612b82af97ec2d.`root`, import SUB:DOLLARa28383612b82af97ec2d.`dist`, import SUB:DOLLARa28383612b82af97ec2d.`buildDirectory`, import SUB:DOLLARa28383612b82af97ec2d.`mkBin`, import SUB:DOLLARa28383612b82af97ec2d.`mkQuick`, import SUB:DOLLARa28383612b82af97ec2d.`mkPack`, import SUB:DOLLARa28383612b82af97ec2d.`testAll`, import SUB:DOLLARa28383612b82af97ec2d.`intellij`, import SUB:DOLLARa28383612b82af97ec2d.`intellijFromSample`, import SUB:DOLLARa28383612b82af97ec2d.`intellijToSample`, import SUB:DOLLARa28383612b82af97ec2d.`pomDependencyExclusions`"
 type="JAVA_MODULE" version="4">
  <component name="NewModuleRootManager" LANGUAGE_LEVEL="JDK_16">
```

The `type=JAVA_MODULE` part seems to be preventing the settings UI from working. The data looks good when collated by `BspProjectResolver`, but something odd is happening after that point.

**Comment by an-tex** [comment]
> While the "root-build" project is imported, and some highlighting of the `./*.sbt` is working, it seems that the `root-build` module does not have the `./project/*.scala` files included as sources nor does it have the UI in the Project Structure dialog to show the SBT settings editor.

For me it shows the sbt tab in the root-build, but I have the same issue regarding missing imports from `./project/*.scala` on the latest IDEA 2021.2, Scala Plugin 2021.2.15

**Comment by jastice** [comment]
@retronym Thanks, got around to reproducing. Followup issue: https://youtrack.jetbrains.com/issue/SCL-19425

**Pull request merged**